tests/e2e/steps_permissionless_ics.go

```diff
@@ -0,0 +1,213 @@
+					InitialHeight: clienttypes.Height{RevisionNumber: 0, RevisionHeight: 1},
+					TopN:          0,
+				},
+				State: State{},
+			},
+			{
+				Action: OptInAction{
+					Chain:     ChainID("cons2"),
+					Validator: ValidatorID("alice"),
+				},
+				State: State{
+					ChainID("provi"): ChainState{
+						HasToValidate: &map[ValidatorID][]ChainID{
+							ValidatorID("alice"): {},
+							ValidatorID("bob"):   {},
+							ValidatorID("carol"): {},
+						},
+					},
+				},
+			},
+		},
+		// Start another permissionless chain with ChainID `consu`
+		// test chain "cons1" is configured with ChainID "consu"
+		stepsStartPermissionlessChain(
+			"cons1", "consu",
+			[]string{"consu", "consu"}, // show up both consumer chains "consu" as proposed chains
+			[]ValidatorID{ValidatorID("bob")}, 0),
+
+		[]Step{
+			{
+				Action: RelayPacketsAction{
+					ChainA:  ChainID("provi"),
+					ChainB:  ChainID("cons1"),
+					Port:    "provider",
+					Channel: 0,
+				},
+				State: State{
+					ChainID("cons1"): ChainState{
+						ValPowers: &map[ValidatorID]uint{
+							ValidatorID("alice"): 0,
+							ValidatorID("bob"):   200,
+							ValidatorID("carol"): 0,
+						},
+					},
+					ChainID("provi"): e2e.ChainState{
+						HasToValidate: &map[ValidatorID][]ChainID{
+							ValidatorID("alice"): {},
+							ValidatorID("bob"):   {"consu"},
+							ValidatorID("carol"): {},
+						},
+					},
+				},
+			},
+		},
+
+		// Test that a validator CANNOT opt-in on a chain with the same ChainID it is already validating
+		[]Step{
+			{
+				Action: OptInAction{
+					Chain:         ChainID("cons2"),
+					Validator:     ValidatorID("alice"),
+					ExpectError:   true,
+					ExpectedError: "already opted in to a chain with the same chain id",
+				},
+				State: State{
+					ChainID("provi"): ChainState{
+						HasToValidate: &map[ValidatorID][]ChainID{
+							ValidatorID("alice"): {},
+							ValidatorID("bob"):   {"consu"},
+							ValidatorID("carol"): {},
+						},
+					},
+					ChainID("cons1"): ChainState{
+						ValPowers: &map[ValidatorID]uint{
+							ValidatorID("alice"): 0, // alice refused to opt in
+							ValidatorID("bob"):   200,
+							ValidatorID("carol"): 0,
+						},
+					},
+				},
+			},
+		},
+		[]Step{
+			{
+				Action: RelayPacketsAction{
+					ChainA:  ChainID("provi"),
+					ChainB:  ChainID("cons1"),
+					Port:    "provider",
+					Channel: 0,
+				},
+				State: State{
+					ChainID("cons1"): ChainState{
+						ValPowers: &map[ValidatorID]uint{
+							ValidatorID("alice"): 0,
+							ValidatorID("bob"):   200,
+							ValidatorID("carol"): 0,
+						},
+					},
+					ChainID("provi"): e2e.ChainState{
+						HasToValidate: &map[ValidatorID][]ChainID{
+							ValidatorID("alice"): {},
+							ValidatorID("bob"):   {"consu"},
+							ValidatorID("carol"): {},
+						},
+					},
+				},
+			},
+		},
+		// test chain hijacking prevention
+		[]Step{
+			// Try to change owner of chain and change deny-/allowlist
```

Comment:
If my understanding is correct, the hijacking prevention is indirectly checked by checking that `bob` is not denylisted. Couldn't we test this by having an `ExpectedError` in `UpdateConsumerChainAction` similarly to `OptInAction`?

Reply:
yes, your understanding is correct. the error check would be more the whitelist approach but possible as well and might just be an additional check beside the one verifying that bob is not denylisted.